### PR TITLE
statistics: fix the problem that the hot cache cannot be emptied when the interval is less than 60

### LINKED ISSUE #4390: The hot cache cannot be cleared, when the interval is less than 60
### What did you do?

stop bench

### What did you expect to see?

hot cache is cleared

### What did you see instead?

there are still many hot peers

### What version of PD are you using (`pd-server -V`)?

v5.0.4

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

close #4390 

### What is changed and how it works?

Our current heartbeat process looks like this.

If the interval is less than the default interval for heartbeats (60 seconds), then we will put it into the cache temporarily and wait until it has made up 60 seconds before considering whether it is hot enough.

For a peer that has just been reported, if the region is also in the hot cache, then there are three cases.
1. the store is the same, then the old item is used directly
2. the store is different, but there is a move peer or transfer leader in the previous round, then the inheritable item is directly used as the old item
3. store is different, and there is no inheritable peer, then choose any peer as old item

The problem occurs in the third branch, if the old peer is used directly without clone, then the old item and the new item will be written at the same time. And when the new peer interval is less than 60 seconds, it means that it will be temporarily put into the cache.

If the old peer will be cooled down at this time, it will keep the peer in the hot cache for a long time and cannot be exited.

This pr does a clone for the third branch to avoid this problem, and changes the store and region heartbeats to configurable in another pr.

### Check List

Tests

- Unit test

### Release note

```release-note
fix the problem that the hot cache cannot be emptied when the interval is less than 60 
```

### CODE REVIEW

**Review by nolouch — COMMENTED** [review]
The issue and PR comments cannot let me understand the problem.

**Review by disksing — APPROVED** [review]

**Comment by rleungx** [comment]
Could you provide more information on how it works?

**Comment by lhy1024** [comment]
@rleungx @nolouch pr desc update

File: server/statistics/kind.go [hunk 1]
```diff
@@ -97,6 +97,27 @@ func (k StoreStatKind) String() string {
...
+	Origin  SourceKind = iota // there is a corresponding peer in this store
+	Inherit                   // there is no a corresponding peer in this store and there is a peer just deleted
+	Adopt                     // there is no a corresponding peer in this store and there is no a peer just deleted, we need to copy from other store
+)
+
+func (k SourceKind) String() string {
+	switch k {
+	case Origin:
+		return "origin"
+	case Inherit:
```
Comment (rleungx) [inline server/statistics/kind.go:113]: What is `Inherit` used for? It looks we only need `Adopt` right now.

File: server/statistics/kind.go [hunk 2]
```diff
@@ -97,6 +97,27 @@ func (k StoreStatKind) String() string {
 	return "unknown StoreStatKind"
 }
 
+// SourceKind represents the statistics item source.
+type SourceKind int
+
+const (
+	Origin  SourceKind = iota // there is a corresponding peer in this store
+	Inherit                   // there is no a corresponding peer in this store and there is a peer just deleted
+	Adopt                     // there is no a corresponding peer in this store and there is no a peer just deleted, we need to copy from other store
```
Comment (rleungx) [inline server/statistics/kind.go:106]:
```suggestion
	Adopt                     // there is no corresponding peer in this store and there is no peer just deleted, we need to copy from other stores
```

File: server/statistics/kind.go [hunk 3]
```diff
@@ -97,6 +97,27 @@ func (k StoreStatKind) String() string {
 	return "unknown StoreStatKind"
 }
 
+// SourceKind represents the statistics item source.
+type SourceKind int
+
+const (
+	Origin  SourceKind = iota // there is a corresponding peer in this store
```
Comment (rleungx) [inline server/statistics/kind.go:104]: seem we don't need to export it.

File: pkg/movingaverage/queue.go [hunk 4]
```diff
@@ -53,3 +53,17 @@ func (sq *SafeQueue) PopFront() interface{} {
 	defer sq.mu.Unlock()
 	return sq.que.PopFront()
 }
+
+func (sq *SafeQueue) Clone() *SafeQueue {
+	sq.mu.Lock()
+	defer sq.mu.Unlock()
+	q := queue.New().Init()
+	for i := 0; i < sq.que.Len(); i++ {
+		node := sq.que.PopFront()
```
Comment (rleungx) [inline pkg/movingaverage/queue.go:62]:
```suggestion
		v := sq.que.PopFront()
```

File: server/statistics/kind.go (hunk 1 above)
Comment (lhy1024) [reply server/statistics/kind.go]: yep, only distinct different source

**Review by rleungx — APPROVED** [review]

**Review by disksing — APPROVED** [review]

**Pull request merged**